enclave/core/sgx/calls.c

```diff
@@ -236,126 +237,225 @@ oe_result_t oe_register_ecall_function_table(
+        {
             OE_RAISE(OE_NOT_FOUND);
+        }
     }
 
     // Fetch matching function.
-    if (args.function_id >= ecall_table.num_ecalls)
+    if (function_id >= ecall_table.num_ecalls)
+    {
         OE_RAISE(OE_NOT_FOUND);
+    }
 
-    func = ecall_table.ecalls[args.function_id];
+    func = ecall_table.ecalls[function_id];
 
-    if (func == NULL)
+    if (NULL == func)
+    {
         OE_RAISE(OE_NOT_FOUND);
+    }
 
     // Allocate buffers in enclave memory
-    buffer = input_buffer = oe_malloc(buffer_size);
-    if (buffer == NULL)
+    buffer = enc_input_buffer = oe_malloc(buffer_size);
+    if (NULL == buffer)
+    {
         OE_RAISE(OE_OUT_OF_MEMORY);
+    }
 
     // Copy input buffer to enclave buffer.
-    memcpy(input_buffer, args.input_buffer, args.input_buffer_size);
+    memcpy(enc_input_buffer, host_input_buffer, input_buffer_size);
 
     // Clear out output buffer.
     // This ensures reproducible behavior if say the function is reading from
     // output buffer.
-    output_buffer = buffer + args.input_buffer_size;
-    memset(output_buffer, 0, args.output_buffer_size);
+    enc_output_buffer = buffer + input_buffer_size;
+    memset(enc_output_buffer, 0, output_buffer_size);
 
     // Call the function.
     func(
-        input_buffer,
-        args.input_buffer_size,
-        output_buffer,
-        args.output_buffer_size,
-        &output_bytes_written);
+        enc_input_buffer,
+        input_buffer_size,
+        enc_output_buffer,
+        output_buffer_size,
+        &enc_output_bytes_written);
 
     // The output_buffer is expected to point to a marshaling struct,
     // whose first field is an oe_result_t. The function is expected
     // to fill this field with the status of the ecall.
-    result = *(oe_result_t*)output_buffer;
+    result = *(oe_result_t*)enc_output_buffer;
 
-    if (result == OE_OK)
+    if (OE_OK == result)
     {
         // Copy outputs to host memory.
-        memcpy(args.output_buffer, output_buffer, output_bytes_written);
+        memcpy(host_output_buffer, enc_output_buffer, enc_output_bytes_written);
 
         // The ecall succeeded.
-        args_ptr->output_bytes_written = output_bytes_written;
-        args_ptr->result = OE_OK;
+        *host_output_bytes_written = enc_output_bytes_written;
+        *host_result = OE_OK;
     }
 
 done:
     if (buffer)
+    {
         oe_free(buffer);
+    }
+
+    return result;
+} /* _handle_ecall_function */
 
+static oe_result_t _handle_call_enclave_function(uint64_t arg_in)
+{
+    oe_call_enclave_function_args_t args, *args_ptr;
+    oe_result_t result = OE_OK;
+
+    // Ensure that args lies outside the enclave.
+    if (!oe_is_outside_enclave(
+            (void*)arg_in, sizeof(oe_call_enclave_function_args_t)))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Copy args to enclave memory to avoid TOCTOU issues.
+    args_ptr = (oe_call_enclave_function_args_t*)arg_in;
+    args = *args_ptr;
```

Comment:
@jxyang With the new version, the call is forwarded to _handle_ecall_function. This implies that the arguments are copied. Can you chime in on whether  the explicit copying in enclave memory in line 385 is needed?
Note: Keep in mind that the compiler could inline any function in the file; but has to preserve C++ semantics.

Reply:
I don't think the copying is necessary. Typically a TOCTOU attack happens in the following way:
```c
func(struct S* p) {
   if (p->f1 != NULL) {
       ... // executing some code, during which p->f1 could be set to NULL  
       a = *p->f1;
   }
}
```
In this case, `oe_is_outside_enclave` is checking `arg_in`, but not individual fields inside it. I don't believe there is risk of TOCTOU even in the original code.

Reply:
removed this copy.